README.md
# measurementProtocol - an R library for sending server-side data to Google Analytics 4

<!-- badges: start -->
[![CRAN](http://www.r-pkg.org/badges/version/measurementProtocol)](https://CRAN.R-project.org/package=measurementProtocol)
![CloudBuild](https://badger-ewjogewawq-ew.a.run.app/build/status?project=mark-edmondson-gde&id=8d88c387-66f1-4ae4-a202-c1cace1fd71f)
[![codecov](https://codecov.io/gh/MarkEdmondson1234/measurementProtocol/branch/master/graph/badge.svg)](https://app.codecov.io/gh/MarkEdmondson1234/measurementProtocol)
<!-- badges: end -->

The [Measurement Protocol v2](https://developers.google.com/analytics/devguides/collection/protocol/ga4) is for sending data to Google Analytics 4.

This library enables the ability to use the API from R.

It also includes functions to use the measurement protocol to track events such as R library loads for package usage analytics.  

Complement the package with [googleAnalyticsR](https://code.markedmondson.me/googleAnalyticsR/) to read the data you send. 

## Install

Install from CRAN

```{r}
install.packages("measurementProtocol")
```

Dev version from GitHub:

```{r}
remotes::install_github("MarkEdmondson1234/measurementProtocol")
```

## Usage 

See the documentation on the `measurementProtocol` website: https://code.markedmondson.me/measurementProtocol/

You need a Google Analytics 4 account to send the hits to.



NEWS.md
# measurementProtocol v0.1.1

* Add Docker image with plumber installed so it can be used with PubSub, available at `gcr.io/gcer-public/measurementProtocol`
* Add `mp_parse_json` that lets you more easily parse generic JSON requests into Measurement Protocol hits, and `mp_parse_gtm` to parse out specifically GA4 events sent to GTM
* Remove unsuppressable start up message "No consent file found" (CRAN check)

# measurementProtocol v0.1.0

* Initial release


tests/testthat.R
library(testthat)
library(measurementProtocol)

test_check("measurementProtocol")


tests/testthat/test_mp.R
test_that("Measurement Protocol Hits",{

  # preferably set this in .Renviron
  Sys.setenv(MP_SECRET="MY_SECRET")

  # your GA4 settings
  my_measurement_id <- "G-43MDXK6CLZ"

  my_connection <- mp_connection(my_measurement_id)
  a_client_id <- 1234567

  event <- mp_event("an_event")
  mp1 <- mp_send(event, a_client_id, my_connection, debug_call = TRUE)
  expect_true(mp1)

  event_error <- mp_event("_an_event")
  mp3 <- mp_send(event_error, a_client_id, my_connection, debug_call = TRUE)
  expect_snapshot(mp3)

  another <- mp_event("another_event")
  mp2 <- mp_send(list(event, another),
             a_client_id, my_connection,
             debug_call = TRUE)
  expect_true(mp2)

  # one item
  it1 <- mp_event_item(item_name = "jeggings",
                   price = 8.88,
                   item_variant = "Black")
  expect_snapshot(it1)

  # many items in a list
  items <- list(
    mp_event_item(item_id = "SKU_12345",
                     price = 9.99,
                     item_brand = "Gucci"),
    it1)
  expect_snapshot(items)

  # construct an event with its own fields
  event1 <- mp_event("add_payment_info",
              params = list(coupon = "SUMMER_FUN",
                            payment_type = "Credit Card",
                            value = 7.77,
                            currency = "USD"),
              items = items)
  expect_snapshot(event1)

  # error item
  expect_error(
    mp_event_item(coupon = "this_will_error")
  )

  expect_equal(class(mp_cid(1)), "character")

  a_trackme_event <- mp_trackme_event("measurementProtocol",
                                      debug_call = TRUE,
                                      say_hello = "hello")
  expect_true(a_trackme_event)

  expect_message(mp_trackme_event("measurementProtocol",
                                  debug_call = TRUE, say_hello = "hello"),
                 "MP Request")

  # custom connection
  custom_connection <- mp_connection(
    my_measurement_id,
    endpoint = "https://gtm.custom.com",
    preview_header = "XXXX"
  )
  expect_snapshot(custom_connection)

  expect_error(
    mp_connection(
      my_measurement_id,
      endpoint = "gtm.custom.com",
      preview_header = "XXXX"
    )
  )

  # error send
  expect_error(
    mp_send(list("error"), client_id = 12324, connection = my_connection)
  )



})


R/utils.R
#' A helper function that tests whether an object is either NULL _or_
#' a list of NULLs
#'
#' @keywords internal
#' @noRd
is.NullOb <- function(x) is.null(x) | all(sapply(x, is.null))

#' Recursively step down into list, removing all such objects
#'
#' @keywords internal
#' @noRd
rmNullObs <- function(x) {
  x <- Filter(Negate(is.NullOb), x)
  lapply(x, function(x) if (is.list(x)) rmNullObs(x) else x)
}


#' Custom message log level
#'
#' @param ... The message(s)
#' @param level The severity
#'
#' @details 0 = everything, 1 = debug, 2=normal, 3=important
#' @keywords internal
#' @noRd
#' @import cli
myMessage <- function(..., level = 2){

  # implement level?

  time <- paste(Sys.time(),">")
  mm <- paste(...)
  if(grepl("^#", mm)){
    cli::cli_h1(mm)
  } else {
    cli::cli_div(theme = list(span.time = list(color = "grey")))
    cli::cli_alert_info("{.time {time}} {mm}")
    cli::cli_end()
  }


}


R/mp_send.R
#' Make a Measurement Protocol v2 request
#'
#' @description
#' `r lifecycle::badge("experimental")`
#' Create a server side call to Google Analytics 4 via its Measurement Protocol
#'
#' @param events The events to send
#' @param client_id The client_id to associate with the event
#' @param connection The connection details created by [mp_connection]
#' @param debug_call Send hits to the Google debug endpoint to validate hits.
#' @param user_id Optional. Unique id for the user
#' @param timestamp_micros Optional. A Unix timestamp (in microseconds) for the time to associate with the event.
#' @param user_properties Optional. The user properties for the measurement sent in as a named list.
#' @param non_personalized_ads Optional. Set to true to indicate these events should not be used for personalized ads.
#'
#' @returns \code{TRUE} if successfully sent the hit.  If \code{debug_call=TRUE} then the JSON response from the debugger endpoint
#'
#' @details
#'
#' Create an API secret via `Admin > Data Streams > choose your stream > Measurement Protocol > Create`
#'
#' To see event parameters, create custom fields in your GA4 account first, to see them in your reports 24hrs after you send them in with this function via `Custom definitions > Create custom dimensions` - `dimension name` will be how it looks like in the reports, `event parameter` will be the parameter you have sent in with the event.
#'
#' `user_id` can be used for cross-platform analysis.
#'
#' `timestamp_micros` should only be set to record events that happened in the past. This value can be overridden via user_property or event timestamps. Events can be backdated up to 48 hours. Note microseconds, not milliseconds.
#'
#' `user_properties` - describe segments of your user base, such as language preference or geographic location.  See [User properties](https://developers.google.com/analytics/devguides/collection/protocol/ga4/user-properties?client_type=gtag)
#'
#' Ensure you also have user permission as specified in the [feature policy](https://developers.google.com/analytics/devguides/collection/protocol/ga4/policy)
#'
#' Invalid events are silently rejected with a 204 response, so use `debug_call=TRUE` to validate your events first.
#'
#' @seealso [Measurement Protocol (Google Analytics 4)](https://developers.google.com/analytics/devguides/collection/protocol/ga4)
#'
#' @export
#' @family Measurement Protocol functions
#' @return `TRUE` if successful, if `debug_call=TRUE` then validation messages if not a valid hit.
#' @examples
#' # preferably set this in .Renviron
#' Sys.setenv(MP_SECRET="MY_SECRET")
#'
#' # your GA4 settings
#' my_measurement_id <- "G-1234"
#'
#' my_connection <- mp_connection(my_measurement_id)
#'
#' a_client_id <- 123.456
#' event <- mp_event("an_event")
#' mp_send(event, a_client_id, my_connection, debug_call = TRUE)
#'
#' # multiple events at same time in a batch
#' another <- mp_event("another_event")
#'
#' mp_send(list(event, another),
#'            a_client_id,
#'            my_connection,
#'            debug_call = TRUE)
#' \dontrun{
#' # you can see sent events in the real-time reports
#' library(googleAnalyticsR)
#' my_property_id <- 206670707
#' ga_data(my_property_id,
#'         dimensions = "eventName",
#'         metrics = "eventCount",
#'         dim_filters = ga_data_filter(
#'            eventName == c("an_event","another_event")),
#'         realtime = TRUE)
#'
#' }
#' @importFrom jsonlite toJSON fromJSON
#' @importFrom httr content POST verbose add_headers
#' @import assertthat
mp_send <- function(
  events,
  client_id,
  connection,
  user_id = NULL,
  debug_call = FALSE,
  timestamp_micros = NULL,
  user_properties = NULL,
  non_personalized_ads = TRUE){

  assert_that(
    is.mp_connection(connection),
    is.flag(debug_call),
    is.flag(non_personalized_ads)
  )

  if(length(events) > 0 &&
     !is.mp_event(events) &&
     !all(unlist(lapply(events, is.mp_event)))){
    stop("Must supply a mp_event object or a list of mp_event objects",
         call. = FALSE)
  }

  # a single event - wrap in a list to make a json array
  if(is.mp_event(events)){
    events <- list(events)
  }

  endpoint <- connection$endpoint
  if(debug_call) endpoint <- "https://www.google-analytics.com/debug/mp/collect"
  my_headers <- connection$preview_header

  the_url <- sprintf(
    "%s?measurement_id=%s&api_secret=%s",
    endpoint, connection$measurement_id, connection$api_secret
  )

  the_body <- rmNullObs(list(
    client_id = as.character(client_id),
    user_id = as.character(user_id),
    timestamp_micros = timestamp_micros,
    user_properties = user_properties,
    non_personalized_ads = non_personalized_ads,
    events = events
  ))

  my_verbose <- NULL
  if(debug_call){
    myMessage("MP Request:",
              the_url,"\n",
              toJSON(the_body,
                     auto_unbox = TRUE,
                     pretty = TRUE),
              level = 3)

    my_verbose <- verbose()

  }

  res <- POST(
    the_url,
    body = the_body,
    connection$preview_header,
    my_verbose,
    encode = "json"
  )

  myMessage("Response: ", res$status, level = 3)

  parsed <- content(res, as = "text", encoding = "UTF-8")

  if(nzchar(parsed) && debug_call){
    o <- fromJSON(parsed)
    if(length(o$validationMessages) > 0) return(o$validationMessages)
    myMessage("No validation messages found", level = 3)
  }

  TRUE
}

#' Create a connection for Measurement Protocol v2
#'
#' Use [mp_connection] to set up the Measurement Protocol connections to pass to [mp_send].  If using Google Tag Manager Server-Side, you can also set up a custom endpoint.
#'
#' @param api_secret The secret generated in the GA4 UI - by default will look for environment arg `MP_SECRET`
#' @param measurement_id The measurement ID associated with a stream
#' @param endpoint If NULL will use Google default, otherwise set to the URL of your Measurement Protocol custom endpoint
#' @param preview_header Only needed for custom endpoints. The `X-Gtm-Server-Preview` HTTP Header found in your GTM debugger
#'
#' @returns An \code{mp_connection} class object
#'
#' @export
#' @examples
#'
#' # custom GTM server side endpoint
#' my_custom_connection <- mp_connection(
#'    my_measurement_id,
#'    endpoint = "https://gtm.example.com",
#'    preview_header = "ZW52LTV8OWdPOExNWFkYjA0Njk4NmQ="
#'  )
#'
#' @rdname mp_send
#' @import assertthat
mp_connection <- function(measurement_id,
                          api_secret = Sys.getenv("MP_SECRET"),
                          endpoint = NULL,
                          preview_header = NULL){
  assert_that(
    is.string(measurement_id),
    is.string(api_secret),
    nzchar(api_secret)
  )

  the_endpoint <- "https://www.google-analytics.com/mp/collect"
  my_headers <- NULL

  if(!is.null(endpoint)){
    assert_that(
      is.string(endpoint),
      grepl("^http", endpoint, ignore.case = TRUE))

    the_endpoint <- endpoint
  }

  if(!is.null(preview_header)){
    assert_that(
      is.string(preview_header),
      nzchar(preview_header))

    my_headers <- add_headers("X-Gtm-Server-Preview" = preview_header)
  }

  structure(list(
    measurement_id = measurement_id,
    api_secret = api_secret,
    endpoint = the_endpoint,
    preview_header = my_headers
  ),
  class = "mp_connection")
}

is.mp_connection <- function(x){
  inherits(x, "mp_connection")
}

#' @export
print.mp_connection <- function(x, ...){
  cat("\n==GA4 MP Connection\n")
  cat("Measurement ID:  ", x$measurement_id)
  cat("\nAPI Secret:       <OBSCURED in .Renviron MP_SECRET>")
  cat("\nEndpoint:        ", x$endpoint)
}


R/mp_trackme.R
.trackme <- new.env()
.trackme$measurement_id <- "G-43MDXK6CLZ"
.trackme$api <- "_hS_7VJARhqbCq9mF3oiNg"

#' Tracking opt-in for this package
#'
#' This is the opt-in function for this package, using [mp_trackme]
#' @returns No return value, called for side effects
#' @export
mp_opt_in <- function(){ # nocov start
  mp_trackme("measurementProtocol")
} # nocov end

get_trackme_file <- function(package){
  file.path(rappdirs::user_config_dir(package), "optin-tracking")
}

#' Opt in or out of package usage tracking
#'
#' You can opt-in or out to sending a measurement protocol hit when you load the package for use in the package's statistics via this function.  No personal data is collected.
#'
#' @export
#' @rdname mp_trackme_event
#' @param package The package you want to track
#'
#'
#' @examples
#'
#' # control your tracking choices via a menu if in interactive session
#' if(interactive()){
#'   mp_trackme()
#' }
#' @import assertthat
#' @importFrom cli cli_h1 cli_alert_info
#' @importFrom utils menu
mp_trackme <- function(package){ # nocov start

  assert_that(is.string(package), nzchar(package))

  cli_h1("Tracking Consent for {package} usage")
  cli_alert_info("This function opts you in or out of sending a tracking hit each time the library loads.")

  opt_in <- menu(
    title = sprintf("Do you opt in to tracking each time you load %s?", package),
    choices = c("Yes","No")
  )

  answer <- FALSE
  if(opt_in == 1){
    answer <- TRUE
  }

  the_file <- get_trackme_file(package)

  if(answer){
    # opt-in
    if(file.exists(the_file)){
      cli::cli_alert_info("Opt-in file {the_file} already found - no more action needed")
      return(invisible(NULL))
    }

    cli::cli_alert_success(
      "Thanks! Your consent and an ID will be marked by the presence of the file {the_file}.  Delete it or run this function again to remove consent.")

    dir.create(rappdirs::user_config_dir(package), showWarnings = FALSE)
    file.create(the_file)
    cid <- mp_cid()
    write(cid, the_file)

    return(invisible(NULL))
  }

  # opt-out

  if(file.exists(the_file)){
    cli::cli_alert_info("Found {the_file} - deleting as you have opted-out")
    unlink(the_file)
  }

  cli::cli_alert_info(
    "No worries! If you change your mind run this function again.")
} # nocov end

#' Package tracking opt-in start-up message
#'
#' Place in .onAttach to have a message appear on first start-up
#' @param package The package to set-up tracking for
#' @param opt_in_function The name of the function for a user to opt-in
#' @returns No return value, called for side effects
#' @import assertthat
#' @noRd
mp_trackme_startup <- function(package, opt_in_function){ # nocov start

  if(is.null(opt_in_function)){
    opt_in_function <- "measurementProtocol::mp_trackme"
  }

  assert_that(
    is.string(package), nzchar(package),
    is.string(opt_in_function)
  )

  if(!file.exists(get_trackme_file(package)) && interactive()){
    packageStartupMessage(
      sprintf("You can opt in to tracking of your use of %s - see ?%s for details.",
              package, opt_in_function)
    )
  }
} # nocov end

#' Send a tracking hit for R package statistics
#'
#' If you opt in, this is the function that fires.  You can use `debug_call=TRUE` to see what would be sent before opting in or out.
#'
#' Running this function will send a Measurement Protocol hit via [mp_send] only if the cache file is present
#' @param package The package name
#' @param opt_in_function The name of the function for a user to opt-in
#' @param debug_call Set as a debug event to see what would be sent
#' @param say_hello If you want to add your own custom message to the event sent, add it here!
#' @returns No return value, called for side effects
#' @export
#'
#' @examples
#'
#' # this only works with a valid opt-in file present
#' mp_trackme_event("googleAnalyticsR")
#'
#' # see what data is sent
#' mp_trackme_event("googleAnalyticsR", debug_call=TRUE)
#'
#' # add your own message!
#' mp_trackme_event("googleAnalyticsR",
#'                  debug_call = TRUE,
#'                  say_hello = "err hello Mark")
#'
#' # placed in .onAttach with function name
#' .onAttach <- function(libname, pkgname){
#'   measurementProtocol::mp_trackme_event(pkgname, opt_in_function = "mp_opt_in")
#'  }
#'
#' @import assertthat
mp_trackme_event <- function(package,
                             debug_call = FALSE,
                             say_hello = NULL,
                             opt_in_function = NULL){

  # extra cautious as this function can prevent package load
  tryCatch(
    trackme_event(package = package,
                  debug_call = debug_call,
                  say_hello = say_hello,
                  opt_in_function = opt_in_function),
    error = function(err){
      warning("Error on package load with mp_trackme_event: ", err$message)
      NULL
    }
  )

}

#' @import assertthat
trackme_event <- function(package,
                          debug_call = FALSE,
                          say_hello = NULL,
                          opt_in_function = NULL){
  assert_that(
    is.string(package), nzchar(package)
  )

  if(!is.null(say_hello)){
    assert_that(is.string(say_hello))
  }

  the_file <- get_trackme_file(package)
  if(!file.exists(the_file) & !debug_call){
    mp_trackme_startup(package, opt_in_function = opt_in_function)
    return(FALSE)
  }

  ss <- utils::sessionInfo()
  event <- mp_event(
    "r_package_loaded",
    params = list(
      r_version = ss$R.version$version.string,
      r_platform = ss$platform,
      r_locale = ss$locale,
      r_system = ss$running,
      say_hello = say_hello,
      package = paste(package,  utils::packageVersion(package))
    )
  )

  if(debug_call){
    cid <- tryCatch(cid <- readLines(the_file)[[1]],
                    error = function(e) "12345678.987654")
  } else {
    cid <- readLines(the_file)[[1]]
  }


  if(any(
    is.null(.trackme$measurement_id),
    is.null(.trackme$api))
  ){
    myMessage("No tracking parameters found, setting dummy values", level = 3)
    m_id = "Measurement_ID"
    api = "API_secret"
  } else {
    m_id = .trackme$measurement_id
    api = .trackme$api
  }

  my_conn <- mp_connection(measurement_id = m_id, api_secret = api)

  if(debug_call){
    return(mp_send(event, client_id = cid,
                   connection = my_conn,
                   debug_call = TRUE))
  } # nocov start
  suppressMessages(
    mp_send(event, client_id = cid,
            connection = my_conn,
            debug_call = debug_call)
  )

  cli::cli_alert_success("Sent library load tracking event")
  # nocov end
}


R/mp_parse_json.R
#' Parse out objects into the Measurement Protocol v2 format for sending
#'
#' This function helps take HTTP events and rearranges its structure so it will work in a MP measurement protocol hit.  This enables HTTP events from say Pub/Sub to be translated into MP hits.
#'
#' @param json The location of a json file or a json string or an R list that has been parsed from json via `jsonlite::fromJSON`
#' @param name_f The function that extracts the event name out of `json`
#' @param params_f An optional function that extracts parameters for the event from `json`
#' @param items_f An optional function that extracts e-commerce items from `json`. Must return a [mp_event_item] object.  you may not need this if the `params_f` includes parsing of e-commerce items
#'
#' @param client_id_f An optional function to extract the client.id.  You will need to supply cid though if using downstream in `mp_send` so it usually is necessary
#' @param user_id_f Optionally include a function that will parse out user_id
#' @param user_properties_f Optionally include a function that will parse out user properties
#'
#' @returns An `mp_parse_json` object that is a list of an `mp_event` object, and `user` fields including client.id, user.id and user properties
#' @export
#' @import jsonlite
#' @import assertthat
#'
#' @details
#'
#' The passed in functions should return NULL if they don't find any entries
#' @examples
#'
#' demo_json <- system.file("example", "pubsub-ga4.json", package = "measurementProtocol")
#' demo_list <- jsonlite::fromJSON(demo_json)
#'
#'
#' # extract the event_name
#' name_f <- function(x) x[["event_name"]]
#'
#' # extract client_id
#' client_id_f <- function(x) x[["client_id"]]
#'
#' # extract user_id
#' user_id_f <- function(x) x[["user_id"]]
#'
#' # simple event
#' mp_parse_json(demo_list,
#'               name_f,
#'               client_id_f = client_id_f,
#'               user_id_f = user_id_f)
#'
#' # params could be assumed to be everything not a event_name of client_id
#' # also not allowed any starting with reserved 'ga_'
#' params_f <- function(x){
#'   x_names <- names(x)[grepl("^x-", names(x))]
#'   ga_names <- names(x)[grepl("^ga_", names(x))]
#'   x[setdiff(names(x), c("client_id","user_id" ,"event_name", x_names, ga_names))]
#'   }
#'
#' # parse including params (could include items as well)
#' parsed_event <- mp_parse_json(demo_list,
#'                               name_f,
#'                               params_f = params_f,
#'                               client_id_f = client_id_f,
#'                               user_id_f = user_id_f)
#' parsed_event
#'
#' # sending to a debug endpoint
#' # preferably set this in .Renviron
#' Sys.setenv(MP_SECRET="MY_SECRET")
#'
#' # replace with your GA4 settings
#' my_measurement_id <- "G-1234"
#' my_connection <- mp_connection(my_measurement_id)
#' mp_send(parsed_event$mp_event,
#'         client_id = parsed_event$user$client_id,
#'         user_id = parsed_event$user$user_id,
#'         user_properties = parsed_event$user$user_properties,
#'         connection = my_connection,
#'         debug_call = TRUE)
#'
#'
mp_parse_json <- function(json,
                          name_f,
                          params_f = NULL,
                          items_f = NULL,
                          client_id_f = NULL,
                          user_id_f = NULL,
                          user_properties_f = NULL
                          ){

  if(!is.list(json)){
    json <- jsonlite::fromJSON(json, simplifyVector = FALSE)
  }

  myMessage("Received JSON:",
            jsonlite::toJSON(json, pretty = TRUE, auto_unbox = TRUE),
            level = 2)

  assert_that(is.function(name_f))

  name <- name_f(json)
  assert_that(is.string(name))

  params <- NULL
  items <- NULL
  client_id <- NULL
  user_id <- NULL
  user_properties <- NULL

  if(!is.null(params_f)){
    params <- params_f(json)
    if(!is.null(params)) assert_that(is.list(params))
  }

  if(!is.null(items_f)){
    items <- items_f(json)
    if(!is.null(items)){
      item_checks <- unlist(lapply(items, is.mp_event_item))
      assert_that(all(item_checks))
    }

  }

  if(!is.null(client_id_f)){
    client_id <-client_id_f(json)

    if(!is.null(client_id)) assert_that(is.string(client_id))
  }

  if(!is.null(user_id_f)){
    user_id <- user_id_f(json)
    if(!is.null(user_id)) assert_that(is.string(user_id))
  }

  if(!is.null(user_properties_f)){
    user_properties <- user_properties_f(json)
    if(!is.null(user_properties)) assert_that(is.list(user_properties))
  }

  structure(list(
    raw = json,
    mp_event = mp_event(
      name = name,
      params = params,
      items = items
    ),
    user = list(
      client_id = client_id,
      user_id = user_id,
      user_properties = user_properties
    )
  ), class = c("mp_parse_json","list"))

}

#' @rdname mp_parse_json
#' @export
#' @examples
#'
#' # mp_parse_gtm internally uses functions demonstrated with mp_parse_json
#' pubsub_event <- mp_parse_gtm(demo_json)
#'
#' mp_send(pubsub_event$mp_event,
#'         client_id = pubsub_event$user$client_id,
#'         user_id = pubsub_event$user$user_id,
#'         user_properties = pubsub_event$user$user_properties,
#'         connection = my_connection,
#'         debug_call = TRUE)
mp_parse_gtm <- function(json){

  name_f <- function(x) x[["event_name"]]
  client_id_f <- function(x) x[["client_id"]]
  user_id_f <- function(x) x[["user_id"]]

  # params could be assumed to be everything not a event_name of client_id
  # also not allowed any starting with reserved 'ga_'
  params_f <- function(x){
    x_names <- names(x)[grepl("^x-", names(x))]
    ga_names <- names(x)[grepl("^ga_", names(x))]
    x[setdiff(names(x), c("client_id","user_id" ,"event_name", x_names, ga_names))]
  }

  mp_parse_json(json,
                name_f,
                params_f = params_f,
                client_id_f = client_id_f,
                user_id_f = user_id_f)

}

#' Parse Pub/Sub responses in plumber files
#'
#' @param pubsub_body The req$postBody of a plumber request
#'
#' @return The Pub/Sub message "data" attribute unencoded into a json string
#' @export
#' @rdname mp_parse_json
#' @examples
#'
#' \dontrun{
#'
#' #* Send forward a measurement protocol hit
#' #* @post /gtm
#' #* @serializer unboxedJSON
#' #* @parser json
#' function(req, res, ga_id) {
#'
#'   pubsub_data <- mp_pubsub_parse(req$postBody)
#'
#'   parsed <- mp_parse_gtm(pubsub_data)
#'
#'   my_connection <- mp_connection(ga_id)
#'
#'   mp_send(parsed$mp_event,
#'           client_id = parsed$user$client_id,
#'           user_id = parsed$user$user_id,
#'           user_properties = parsed$user$user_properties,
#'           connection = my_connection)
#'
#'   "OK"
#'   }
#'
#'
#' }
mp_pubsub <- function(pubsub_body){
  pubsub_data <- NULL
  if(!is.null(pubsub_body) && nzchar(pubsub_body)){
    pubsub_data <- jsonlite::fromJSON(pubsub_body)
  }

  if(is.null(pubsub_data$message) ||
     is.null(pubsub_data$message$data)){
    message("Pub/Sub Message Data was invalid")
    return(FALSE)
  }

  message <- pubsub_data$message

  cat(as.character(Sys.time()),
      "-pubsub-message_id-",
      message$message_id,"-",
      message$publish_time,"-\n")

  rawToChar(jsonlite::base64_dec(message$data))

}


inst/plumber/api.R
library(measurementProtocol)

# get mp_secret from env MP_SECRET

#* Home page
#* @get /
#* @serializer html
function() {
  "<html>
    <h1>measurementProtocol + plumber</h1>
    <p>POST to /gtm?ga_id=G-123456<p>
    <p>Debug via /gtm?ga_id=G-123456&debug=1<p>
    <p><a href=/__docs__/>Swagger docs</a>
  </html>"
}

# For parsing Pub/Sub messages
mp_pubsub <- function(pubsub_body){
  pubsub_data <- NULL
  if(!is.null(pubsub_body) && nzchar(pubsub_body)){
    pubsub_data <- jsonlite::fromJSON(pubsub_body)
  }

  if(is.null(pubsub_data$message) ||
     is.null(pubsub_data$message$data)){
    message("Pub/Sub Message Data was invalid")
    return(FALSE)
  }

  message <- pubsub_data$message

  cat(as.character(Sys.time()),
      "-pubsub-message_id-",
      message$message_id,"-",
      message$publish_time,"-\n")

  rawToChar(jsonlite::base64_dec(message$data))

}

#* Send forward a measurement protocol hit
#* @post /gtm
#* @serializer unboxedJSON
#* @parser json
function(req, res, ga_id, debug = 0) {

  pubsub_data <- mp_pubsub(req$postBody)

  if(isFALSE(pubsub_data)){
    res$status <- 400
    return(list(error="Pub/Sub data parsing error"))
  }

  parsed <- suppressMessages(
    mp_parse_gtm(pubsub_data)
    )

  my_connection <- mp_connection(ga_id)

  message(
    sprintf("Sending event: %s for client.id %s",
            parsed$mp_event$name,
            parsed$user$client_id)
  )

  sent <- mp_send(parsed$mp_event,
                  client_id = parsed$user$client_id,
                  user_id = parsed$user$user_id,
                  user_properties = parsed$user$user_properties,
                  connection = my_connection,
                  debug_call = if(debug != 0) TRUE else FALSE)

  if(!isTRUE(sent)){
    res$status <- 400
    return(list(error="MP hit failed to send"))
  }

  "OK"
}


cloud_build/builds.R
library(googleCloudRunner)

repo <- "MarkEdmondson1234/measurementProtocol"

cr_deploy_pkgdown(repo,
                  secret = "github-ssh",
                  cloudbuild_file = "cloud_build/pkgdown.yml")

cr_deploy_packagetests(
  cloudbuild_file = "cloud_build/testthat.yml",
  trigger_repo = cr_buildtrigger_repo(repo,branch = "^master$")
)


cr_deploy_docker_trigger(
  repo = cr_buildtrigger_repo(repo),
  image = "measurementProtocol", # forced down to be lowercase
  projectId_target = "gcer-public",
  timeout = 3600
)

cr_deploy_plumber(
  "inst/plumber",
  remote = "measurement_protocol_proxy",
  env_vars = paste0("MP_SECRET=", Sys.getenv("MP_SECRET"))
)

# curl -X POST "https://measurement-protocol-proxy-ewjogewawq-ew.a.run.app/gtm?gtm_id=dfdsfsf" -H "accept: application/json" -d '{"event_name":"hi","client_id":"1234"}'


bs <- c(
  cr_buildstep_docker(
    "measurement-protocol-proxy",
    kaniko_cache = TRUE,
    dir = "inst/plumber"
  ),
  cr_buildstep_run(
    "measurement-protocol-proxy",
    image = "gcr.io/$PROJECT_ID/measurement-protocol-proxy:$BUILD_ID",
    env_vars = paste0("MP_SECRET=", Sys.getenv("MP_SECRET"))
  )
)

yaml <- cr_build_yaml(bs)
build <- cr_build_make(yaml)

cr_buildtrigger(
  build,
  name = "measurement-protocol-proxy-build",
  trigger = cr_buildtrigger_repo(repo),
  includedFiles = "inst/plumber/**",
  overwrite = TRUE
)




R/mp_event.R
#' Create a Measurement Protocol Event
#'
#' @description
#' `r lifecycle::badge("experimental")`
#' This creates an event to send via [mp_send]
#'
#' @param name The event name to send in
#' @param params Optional event parameters sent in as a named list
#' @param items Optional items created via [mp_event_item]
#'
#' @returns An `mp_event` object
#'
#' @export
#' @family Measurement Protocol functions
#' @import assertthat
#' @examples
#'
#' mp_event("custom_event")
#' mp_event("custom_event", params = list(my_param = "SUPER"))
mp_event <- function(name, params = NULL, items = NULL){

  if(!is.null(items)){
    item_checks <- unlist(lapply(items, is.mp_event_item))
    assert_that(all(item_checks))

    params <- c(params, list(items = items))
  }

  structure(
    rmNullObs(list(
      name = name,
      params = params
    )), class = c("mp_event","list")
  )
}

is.mp_event <- function(x){
  inherits(x, "mp_event")
}

#' @export
print.mp_event <- function(x, ...){
  cat("\n==GA4 MP Event\n")
  print(jsonlite::toJSON(x, pretty = TRUE, auto_unbox = TRUE))
}

#' Create an Measurement Protocol Item Property for an Event
#'
#' @description
#' `r lifecycle::badge("experimental")`
#' Some events work with item properties
#'
#' @param item_id Item ID
#' @param item_name Item Name
#' @param coupon Coupon
#' @param discount Discount
#' @param affiliation Affiliation
#' @param item_brand Brand
#' @param item_category Category
#' @param item_variant Variant
#' @param price Price
#' @param currency Currency
#'
#' @returns An `mp_event_item` object
#'
#' @export
#' @family Measurement Protocol functions
#' @examples
#' # one item
#' mp_event_item(item_name = "jeggings",
#'                  price = 8.88,
#'                  item_variant = "Black")
#'
#' # many items in a list
#' items <- list(
#'   mp_event_item(item_id = "SKU_12345",
#'                 price = 9.99,
#'                 item_brand = "Gucci"),
#'   mp_event_item(item_name = "jeggings",
#'                 price = 8.88,
#'                 item_variant = "Black"))
#'
#' # construct an event with its own fields
#' mp_event("add_payment_info",
#'          params = list(coupon = "SUMMER_FUN",
#'                        payment_type = "Credit Card",
#'                        value = 7.77,
#'                        currency = "USD"),
#'          items = items)
#'
mp_event_item <- function(
  item_id = NULL,
  item_name = NULL,
  coupon = NULL,
  discount = NULL,
  affiliation = NULL,
  item_brand = NULL,
  item_category = NULL,
  item_variant = NULL,
  price = NULL,
  currency = NULL
){

  if(all(is.null(item_id), is.null(item_name))){
    stop("One of item_id or item_name is required", call. = FALSE)
  }

  structure(
    rmNullObs(list(
      item_id = item_id,
      item_name = item_name,
      coupon = coupon,
      discount = discount,
      affiliation = affiliation,
      item_brand = item_brand,
      item_category = item_category,
      item_variant = item_variant,
      price = price,
      currency = currency
    )), class = c("mp_event_item","list")
  )

}

is.mp_event_item <- function(x){
  inherits(x, "mp_event_item")
}

#' @export
print.mp_event_item <- function(x, ...){
  cat("\n==GA4 MP Event Item\n")
  print(jsonlite::toJSON(x, pretty = TRUE, auto_unbox = TRUE))
}

#' Generate a random client_id
#'
#' This has a random number plus a timestamp
#'
#' @param seed If you set a seed, then the random number will be the same for each value
#' @returns A string suitable as an Id with a random number plus a timestamp delimited by a period.
#' @export
#' @family Measurement Protocol functions
#' @examples
#'
#' # random Id
#' mp_cid()
#'
#' # fix the random number (but not the timestamp)
#' mp_cid(1)
mp_cid <- function(seed = NULL){

  if(!is.null(seed)){
    set.seed(seed)
  }
  rand <- round(stats::runif(1, min = 1, max = 100000000))
  ts <- round(as.numeric(Sys.time()))

  paste0(rand,".",ts)

}
